### PR TITLE
Fix #3666, plus some drive-by edits

### LINKED ISSUE #3666: take_while_view::sentinel declaration is missing the Pred parameter
We have

```
namespace std::ranges {
  template<view V, class Pred>
  requires input_range<V> && is_object_v<Pred> &&
  indirect_unary_predicate<const Pred, iterator_t<V>>
class take_while_view
```

but

```
namespace std::ranges {
    template<class V>
    template<bool Const>
    class take_while_view<V>::sentinel {
```

(missing the second template argument and the requires-clause).

### PR DESCRIPTION
**Pull request by CaseyCarter**
See the three commit logs.

Note that the change to `basic_istream_view::iterator` uses the name `default_initializable` instead of `default_constructible` in anticipation of LWG motion 2.

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -6159,7 +6187,9 @@
 \indexlibraryglobal{elements_view::iterator}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V, size_t N>
+  template<input_range V, size_t N>
+    requires view<V> && @\placeholder{has-tuple-element}@<range_value_t<V>, N> &&
+      @\placeholder{has-tuple-element}@<remove_reference_t<range_reference_t<V>>, N>
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
             @\placeholder{has-tuple-element}@<remove_reference_t<range_reference_t<V>>, N>
```

Comment (CaseyCarter) [reply source/ranges.tex]: Thanks - fixed.

File: source/ranges.tex [hunk 2]
```diff
@@ -213,9 +213,10 @@
   namespace views { inline constexpr @\unspec@ take = @\unspec@; }
 
   // \ref{range.take.while}, take while view
-  template<view V, class Pred>
-    requires input_range<V> && is_object_v<Pred> &&
-             indirect_unary_predicate<const Pred, iterator_t<V>>
+
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
```

File: source/ranges.tex [hunk 3]
```diff
@@ -227,9 +228,10 @@
   namespace views { inline constexpr @\unspec@ drop = @\unspec@; }
 
   // \ref{range.drop.while}, drop while view
-  template<view V, class Pred>
-    requires input_range<V> && is_object_v<Pred> &&
-             indirect_unary_predicate<const Pred, iterator_t<V>>
+
```
Comment (johelegp) [inline source/ranges.tex]:
```suggestion
```

File: source/ranges.tex [hunk 4]
```diff
@@ -2270,7 +2272,8 @@
 \indexlibraryglobal{iota_view::iterator}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class W, class Bound>
+  template<weakly_incrementable W, semiregular Bound>
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{weakly_incrementable}@ W, @\libconcept{semiregular}@ Bound>
```

File: source/ranges.tex [hunk 5]
```diff
@@ -2640,7 +2643,8 @@
 \indexlibraryglobal{iota_view::sentinel}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class W, class Bound>
+  template<weakly_incrementable W, semiregular Bound>
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{weakly_incrementable}@ W, @\libconcept{semiregular}@ Bound>
```

File: source/ranges.tex [hunk 6]
```diff
@@ -3221,7 +3227,8 @@
 \indexlibrarymember{iterator}{filter_view}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V, class Pred>
+  template<input_range V, indirect_unary_predicate<iterator_t<V>> Pred>
+    requires view<V> && is_object_v<Pred>
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{input_range}@ V, @\libconcept{indirect_unary_predicate}@<iterator_t<V>> Pred>
    requires @\libconcept{view}@<V> && is_object_v<Pred>
```

File: source/ranges.tex [hunk 7]
```diff
@@ -3471,7 +3478,8 @@
 \indexlibrarymember{sentinel}{filter_view}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V, class Pred>
+  template<input_range V, indirect_unary_predicate<iterator_t<V>> Pred>
+    requires view<V> && is_object_v<Pred>
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{input_range}@ V, @\libconcept{indirect_unary_predicate}@<iterator_t<V>> Pred>
    requires @\libconcept{view}@<V> && is_object_v<Pred>
```

File: source/ranges.tex [hunk 8]
```diff
@@ -3704,7 +3712,9 @@
 \indexlibraryglobal{transform_view::iterator}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V, class F>
+  template<input_range V, copy_constructible F>
+    requires view<V> && is_object_v<F> &&
+             regular_invocable<F&, range_reference_t<V>>
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{input_range}@ V, @\libconcept{copy_constructible}@ F>
    requires @\libconcept{view}@<V> && is_object_v<F> &&
             @\libconcept{regular_invocable}@<F&, range_reference_t<V>>
```

File: source/ranges.tex [hunk 9]
```diff
@@ -4107,7 +4117,9 @@
 \indexlibraryglobal{transform_view::sentinel}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V, class F>
+  template<input_range V, copy_constructible F>
+    requires view<V> && is_object_v<F> &&
+             regular_invocable<F&, range_reference_t<V>>
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{input_range}@ V, @\libconcept{copy_constructible}@ F>
    requires @\libconcept{view}@<V> && is_object_v<F> &&
             @\libconcept{regular_invocable}@<F&, range_reference_t<V>>
```

File: source/ranges.tex [hunk 10]
```diff
@@ -4361,7 +4373,7 @@
 \indexlibraryglobal{take_view::sentinel}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V>
+  template<view V>
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{view}@ V>
```

File: source/ranges.tex [hunk 11]
```diff
@@ -4532,7 +4544,9 @@
 \indexlibraryglobal{take_while_view::sentinel}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V, class Pred>
+  template<view V, class Pred>
+    requires input_range<V> && is_object_v<Pred> &&
+             indirect_unary_predicate<const Pred, iterator_t<V>>
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{view}@ V, class Pred>
    requires @\libconcept{input_range}@<V> && is_object_v<Pred> &&
             @\libconcept{indirect_unary_predicate}@<const Pred, iterator_t<V>>
```

File: source/ranges.tex [hunk 12]
```diff
@@ -4965,7 +4979,10 @@
 \indexlibraryglobal{join_view::iterator}%
 \begin{codeblock}
 namespace std::ranges {
-template<class V>
+  template<input_range V>
+    requires view<V> && input_range<range_reference_t<V>> &&
+             (is_reference_v<range_reference_t<V>> ||
+              view<range_value_t<V>>)
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{input_range}@ V>
    requires @\libconcept{view}@<V> && @\libconcept{input_range}@<range_reference_t<V>> &&
             (is_reference_v<range_reference_t<V>> ||
              @\libconcept{view}@<range_value_t<V>>)
```

File: source/ranges.tex [hunk 13]
```diff
@@ -5274,7 +5291,10 @@
 \indexlibraryglobal{join_view::sentinel}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V>
+  template<input_range V>
+    requires view<V> && input_range<range_reference_t<V>> &&
+             (is_reference_v<range_reference_t<V>> ||
+              view<range_value_t<V>>)
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{input_range}@ V>
    requires @\libconcept{view}@<V> && @\libconcept{input_range}@<range_reference_t<V>> &&
             (is_reference_v<range_reference_t<V>> ||
              @\libconcept{view}@<range_value_t<V>>)
```

File: source/ranges.tex [hunk 14]
```diff
@@ -5466,7 +5486,10 @@
 \indexlibraryglobal{split_view::outer-iterator}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V, class Pattern>
+  template<input_range V, forward_range Pattern>
+    requires view<V> && view<Pattern> &&
+             indirectly_comparable<iterator_t<V>, iterator_t<Pattern>, ranges::equal_to> &&
+             (forward_range<V> || @\exposconcept{tiny-range}@<Pattern>)
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<input_range V, forward_range Pattern>
    requires @\libconcept{view}@<V> && @\libconcept{view}@<Pattern> &&
             @\libconcept{indirectly_comparable}@<iterator_t<V>, iterator_t<Pattern>, ranges::equal_to> &&
             (@\libconcept{forward_range}@<V> || @\exposconcept{tiny-range}@<Pattern>)
```

File: source/ranges.tex [hunk 15]
```diff
@@ -5624,7 +5647,10 @@
 \indexlibraryglobal{split_view::outer-iterator::value_type}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V, class Pattern>
+  template<input_range V, forward_range Pattern>
+    requires view<V> && view<Pattern> &&
+             indirectly_comparable<iterator_t<V>, iterator_t<Pattern>, ranges::equal_to> &&
+             (forward_range<V> || @\exposconcept{tiny-range}@<Pattern>)
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<input_range V, forward_range Pattern>
    requires @\libconcept{view}@<V> && @\libconcept{view}@<Pattern> &&
             @\libconcept{indirectly_comparable}@<iterator_t<V>, iterator_t<Pattern>, ranges::equal_to> &&
             (@\libconcept{forward_range}@<V> || @\exposconcept{tiny-range}@<Pattern>)
```

File: source/ranges.tex [hunk 16]
```diff
@@ -5689,7 +5715,10 @@
 \indexlibraryglobal{split_view::inner-iterator}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V, class Pattern>
+  template<input_range V, forward_range Pattern>
+    requires view<V> && view<Pattern> &&
+             indirectly_comparable<iterator_t<V>, iterator_t<Pattern>, ranges::equal_to> &&
+             (forward_range<V> || @\exposconcept{tiny-range}@<Pattern>)
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<input_range V, forward_range Pattern>
    requires @\libconcept{view}@<V> && @\libconcept{view}@<Pattern> &&
             @\libconcept{indirectly_comparable}@<iterator_t<V>, iterator_t<Pattern>, ranges::equal_to> &&
             (@\libconcept{forward_range}@<V> || @\exposconcept{tiny-range}@<Pattern>)
```

**Comment by zygoloid** [comment]
Manually rebased (and basically recreated from scratch, since the *template-head*s have changed substantially) and merged in ac3fceec6107d67ed69535583ad83e35da9de218.

File: source/ranges.tex [hunk 17]
```diff
@@ -6294,7 +6324,9 @@
 \indexlibraryglobal{elements_view::iterator}%
 \begin{codeblock}
 namespace std::ranges {
-  template<class V, size_t N>
+  template<input_range V, size_t N>
+    requires view<V> && @\placeholder{has-tuple-element}@<range_value_t<V>, N> &&
+             @\placeholder{has-tuple-element}@<remove_reference_t<range_reference_t<V>>, N>
```
Comment (CaseyCarter) [inline source/ranges.tex]:
```suggestion
  template<@\libconcept{input_range}@ V, size_t N>
    requires @\libconcept{view}@<V> && @\exposconcept{has-tuple-element}@<range_value_t<V>, N> &&
             @\exposconcept{has-tuple-element}@<remove_reference_t<range_reference_t<V>>, N>
```

**Pull request closed without merging**